apps/emqx_management/src/emqx_mgmt_api_clients.erl

```diff
@@ -890,6 +945,218 @@ list_clients(QString) ->
             {200, Response}
     end.
 
+list_clients_v2(get, #{query_string := QString0}) ->
+    Nodes = emqx:running_nodes(),
+    case maps:get(<<"cursor">>, QString0, none) of
+        none ->
+            Cursor = initial_ets_cursor(Nodes),
+            do_list_clients_v2(Nodes, Cursor, QString0);
+        CursorBin when is_binary(CursorBin) ->
+            case parse_cursor(CursorBin, Nodes) of
+                {ok, Cursor} ->
+                    do_list_clients_v2(Nodes, Cursor, QString0);
+                {error, bad_cursor} ->
+                    ?BAD_REQUEST(<<"bad cursor">>)
+            end
+    end.
+
+do_list_clients_v2(Nodes, Cursor, QString0) ->
+    Limit = maps:get(<<"limit">>, QString0, 100),
+    Acc = #{
+        rows => [],
+        n => 0,
+        limit => Limit
+    },
+    do_list_clients_v2(Nodes, Cursor, QString0, Acc).
+
+do_list_clients_v2(_Nodes, Cursor = done, _QString, Acc) ->
+    format_results(Acc, Cursor);
+do_list_clients_v2(Nodes, Cursor = #{type := ?CURSOR_TYPE_ETS, node := Node}, QString0, Acc0) ->
+    {Rows, NewCursor} = do_ets_select(Nodes, QString0, Cursor),
+    Acc1 = maps:update_with(rows, fun(Rs) -> [{Node, Rows} | Rs] end, Acc0),
+    Acc = #{limit := Limit, n := N} = maps:update_with(n, fun(N) -> N + length(Rows) end, Acc1),
+    case N >= Limit of
```

Comment:
we should try to compare with `>= Limit + 1` or `> Limit` but do not return the last item in the list (and the last item's key should be the next cursor).
this should eliminate the false positive for `hasnext`

Reply:
I think just the last seen clientid is not enough for a cursor, since we have to i) traverse different nodes which have different first and last clientids; ii) after all ETS tables are traversed, we traverse DS DB, which is not ordered by clientid.  That's why I used the current cursor implementation.

Reply:
On second thought, I think I see now what you meant.  Using just the last clientid is still not enough as a cursor, but we could use that and restart `ets:select` using it instead of holding on to the continuation.  :thinking: